src/lenskit/metrics/_accum.py

```diff
@@ -0,0 +1,77 @@
+# This file is part of LensKit.
+# Copyright (C) 2018-2023 Boise State University.
+# Copyright (C) 2023-2025 Drexel University.
+# Licensed under the MIT license, see LICENSE.md for details.
+# SPDX-License-Identifier: MIT
+
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+
+from lenskit.data import ItemList
+
+from ._base import Metric
+
+
+class MetricAccumulator:
+    """
+    Handles metric registration, accumulation of metric measurements
+    per recommendation list, and aggregation of summary statistics.
+    """
+
+    def __init__(self):
+        self._metrics: List[Metric] = []
+        self._results: List[dict[str, Any]] = []
+
+    def add_metric(self, metric: Metric) -> None:
+        """
+        Register a metric to be accumulated.
+
+        Args:
+            metric: An instance of a Metric subclass.
+        """
+        self._metrics.append(metric)
+
+    def measure_list(
+        self,
+        output: ItemList,
+        test: ItemList,
+        user_id: Optional[Any] = None,
+    ) -> None:
+        """
+        Measure and accumulate metrics for a single recommendation list.
+
+        Args:
+            output: The recommendation items for a user.
+            test: The ground-truth items for the same user.
+            user_id: Optional identifier for the user.
+        """
+        record = {"user_id": user_id}
+        for metric in self._metrics:
+            result = metric.measure_list(output, test)
+            record[metric.label] = metric.extract_list_metrics(result)
```

Comment:
Is this what we want if the list metrics return a dict? Possibly, but need to take a think about it.